### PR TITLE
Improve documentation

### PR DESCRIPTION
**Pull request by kirillbobyrev**
* Fix a bunch of typos
* Improve wording where possible
* Use Markdown's best practices (list alignment, no multiple top level
headings, code snippets should specify language, newlines where
approprriate, etc)
* Reformat where needed to enforce 80 column width limit

### CODE REVIEW

File: config.md [hunk 1]
```diff
@@ -44,19 +44,19 @@ Project configuration applies to files under its tree (`proj/.clangd` configures
 Configuration is combined when this is sensible. In case of conflicts, user
 config has the highest precedence, then inner project, then outer project.
 
-# Schema
+## Schema
```
Comment (sam-mccall) [inline config.md:47]: The problem with bumping Schema to H2, blocks to H3, and attributes within blocks to H4, is we don't have a visually distinct style for H4 headings. Maybe we should! (but definitely its own patch).

Is having a single H1 important somehow?
(I don't have any objection if it looks just as good, but if it looks worse then we need a reason!)

File: config.md [hunk 2]
```diff
@@ -120,25 +120,26 @@ List of flags to remove from the compile command.
...
 Controls how clangd understands code outside the current file.
 
-```
+```yaml
 Index:
   Background: Skip     # Disable slow background indexing of these files.
 ```
 
-clangd's indexes provide information about symbols that isn't available to
+clangd indexes provide information about symbols that isn't available to
```
Comment (sam-mccall) [inline config.md:139]: both are correct, but I prefer "clangd's" here as it's more specific about the relationship

File: design/code.md [hunk 3]
```diff
@@ -81,7 +81,7 @@ to work, particularly on Mac. These live in [CommandMangler].
 ### Diagnostics
 
 During parsing, clang emits diagnostics through a `DiagnosticConsumer` callback.
-clangd's [StoreDiags] implementation converts them into [Diag] objects,
+clangd [StoreDiags] implementation converts them into [Diag] objects,
```
Comment (sam-mccall) [inline design/code.md:84]: I think this is an error, "clangd StoreDiags implementation" is a cluster of nouns I can't parse

File: design/indexing.md [hunk 4]
```diff
@@ -63,10 +63,10 @@ is found, the compile command for each source file is placed on a queue.
 Before indexing each file, the index checks for a cached `*.idx` file on disk.
 After indexing, it writes this file. This avoids reindexing on startup if
 nothing changed since last time.
-These files are located in `.cache/clangd/index/` next to `compile_commands.json`.
-For headers with no CDB, such as the standard library, they are in `clangd/index`
-under the user's cache directory (`$XDG_CACHE_HOME`, `DARWIN_USER_CACHE_DIR`, or
-`%LocalAppData%`).
+These files are located in `.cache/clangd/index/` next to
+`compile_commands.json`. For headers with no CDB, such as the standard
```
Comment (sam-mccall) [inline design/indexing.md:67]: FWIW, I find this kind of rewrapping makes the source code harder to read and edit, and it has no effect on the output. (Analogy: paragraphs are like functions, line breaks in the source are like blank lines in functions, separating thoughts)
But up to you.

File: extensions.md [hunk 5]
```diff
@@ -14,36 +15,39 @@ We try to do this sparingly. The most important considerations are:
...
-{% include toc.md %} 
+{% include toc.md %}
 
 ## Switch between source/header
 {:.v6}
 
 Lets editors switch between the main source file (`*.cpp`) and header (`*.h`).
 
 **New client->server request**: `textDocument/switchSourceHeader`.
-  - Params: `TextDocumentIdentifier`: an open file.
```
Comment (sam-mccall) [inline extensions.md:25]: Hmm, this indented formatting was deliberate - it closely matches the rendering, and visually associates the params/result with the request as opposed to the overall heading.

Is there a reason it needs to be changed?

File: extensions.md [hunk 6]
```diff
@@ -14,36 +15,39 @@ We try to do this sparingly. The most important considerations are:
...
+  provided) or source file (if a header was provided).
 
 If the corresponding file can't be determined, `""` is returned.
 [bug?](https://github.com/clangd/clangd/issues/12)
 
 ## File status
 {:.v8}
 
-Provides information about activity on clangd's per-file worker thread.
+Provides information about activity on clangd per-file worker thread.
```
Comment (sam-mccall) [inline extensions.md:37]: again, this seems ungrammatical without the 's

File: extensions.md [hunk 7]
```diff
@@ -122,49 +136,69 @@ further information (like definition location, which may require consulting an
...
+- `utf-32`: `character` counts codepoints
 
 **New InitializeResponse property**: `offsetEncoding: string`:
-  - Specifies the encoding that was selected by the server, and should be used.
-  - This should be one of the requested encodings, or `"utf-16"` if none ore supported.
-  - Only sent if the client capability was specified (or the equivalent command-line flag `-offset-encoding=utf-8`).
-  
+
+- Specifies the encoding that was selected by the server, and should be used.
+- This should be one of the requested encodings, or `"utf-16"` if none ore
```
Comment (sam-mccall) [inline extensions.md:174]: while here: *are* supported

File: extensions.md [hunk 8]
```diff
@@ -122,49 +136,69 @@ further information (like definition location, which may require consulting an
...
+
 **Advice for clients using this extension**:
-  - clients that only support UTF-8 should send `offsetEncoding: ["utf-8"]` in their client capabilities.
-    This will cause the server to use UTF-8 if supported.
-  - clients that prefer UTF-8 but can use UTF-16 should send `offsetEncoding: ["utf-8", "utf-16"]` and observe the selected encoding in the `InitializeResponse`, defaulting to UTF-16 if it's not present.
-    This will negotiate UTF-8 with servers that support it.
-  - clients that prefer UTF-16 may send `offsetEncoding: ["utf-16"]` or simply not use the extension.
- 
+
+- clients that only support UTF-8 should send `offsetEncoding: ["utf-8"]` in
```
Comment (sam-mccall) [inline extensions.md:181]: these seem to have been rewrapped quite short

(FWIW, when editing markdown, I'm usually in an editor with softwrap on, and typically 90-120 cols, so I find too-short lines more of a problem than too-long ones. I'm sure workflows vary though!)

File: features.md [hunk 9]
```diff
@@ -41,7 +41,6 @@ and style issues.
 clangd respects your project's `.clang-tidy` file which controls the checks to
 run. Not all checks work within clangd.
 
-
```
Comment (sam-mccall) [inline features.md:44]: again, these breaks are deliberate as this page has a lot of sections and subsections and they can be hard to visually distinguish. Do you find the opposite?

File: installation.md [hunk 10]
```diff
@@ -86,21 +91,23 @@ You should see errors highlighted and completions as you type.
...
+- rename symbol (`:YcmCompleter RefactorRename`)
 
 ### Under the hood
 
 - **Debug logs**: run `:YcmDebugInfo` to see clangd status, and `:YcmToggleLogs`
   to view clangd's debug logs.
 - **Command-line flags**: Set `g:ycm_clangd_args` in `.vimrc`, e.g.:
-```
+
+```vim
```
Comment (sam-mccall) [inline installation.md:107]: surely this pulls it out of the bulleted list?

File: troubleshooting.md [hunk 11]
```diff
@@ -14,7 +14,7 @@ pre-release snapshot before you try anything else!
 
 ## Gathering logs
 
-Your editor should make clangd's `stderr` log available (see [installation][]
+Your editor should make clangd `stderr` log available (see [installation][]
```
Comment (sam-mccall) [inline troubleshooting.md:17]: this one is ungrammatical too

**Review by sam-mccall — COMMENTED** [review]
Thanks, and sorry for the long turnaround.

There's a mix of content and source-formatting changes here which I know is tempting... but I think the latter is largely a question of taste.

File: config.md (hunk 1 above)
Comment (kirillbobyrev) [reply config.md:47]: Reverted this back. There is no particular "problem" with multiple H1 but it is a convention to only have one (e.g. HTML `<head>`).

File: extensions.md (hunk 5 above)
Comment (kirillbobyrev) [reply extensions.md:25]: From Markdown point of view this is just equivalent and the indentation usually means actual nested indentation (hence, idiomatic MD). Decreasing indentation level by 1 also saves some space.

I don't have a strong opinion, though, and am OK to revert this, this is simply surprising to me from the aesthetic point of view.

**Comment by kirillbobyrev** [comment]
@sam-mccall I think the overall differences we have in regards to MD documents is that you see them as "mostly readable" text documents that have a somewhat close correspondance with the output while I perceive MD as "HTML-like" documents that are bound to produce nice output but do not have to visually look better. Hence, MD convention (such as "always separate by newline", "have a single columnwidth limit") does not scare me too much (maybe I just got used to reading a lot of MD which looks awkward). I'm not opposed to reverting code for unresolved comments but the point is mostly to make everything coherent and have a unified style (the "default" conventional MD style). WDYT?

**Pull request closed without merging**